src/platform/posix/posix_pollq_kqueue.c

```diff
@@ -0,0 +1,392 @@
+	}
+
+	if (nevents > 0) {
+		int rv = kevent(pq->kq, kevents, nevents, NULL, 0, NULL);
+		NNI_ASSERT(rv >= 0);
+		node->events |= events;
+	}
+
+	nni_mtx_unlock(&pq->mtx);
+}
+
+void
+nni_posix_pollq_disarm(nni_posix_pollq_node *node, int events)
+{
+	struct kevent kevents[2];
+	int           nevents = 0;
+
+	nni_posix_pollq *pq = node->pq;
+	if (pq == NULL) {
+		return;
+	}
+
+	nni_mtx_lock(&pq->mtx);
+
+	if ((node->events & POLLIN) && (events & POLLIN)) {
+		EV_SET(&kevents[nevents++], (uintptr_t) node->fd, EVFILT_READ,
+		    EV_DISABLE, 0, 0, (kevent_udata_t) node);
+	}
+
+	if ((node->events & POLLOUT) && (events & POLLOUT)) {
+		EV_SET(&kevents[nevents++], (uintptr_t) node->fd, EVFILT_WRITE,
+		    EV_DISABLE, 0, 0, (kevent_udata_t) node);
+	}
+
+	if (nevents > 0) {
+		int rv = kevent(pq->kq, kevents, nevents, NULL, 0, NULL);
+		if (rv < 0 && errno != ENOENT && errno != EBADF) {
+			NNI_ASSERT(false);
+		}
+		node->events &= ~events;
+	}
+
+	nni_mtx_unlock(&pq->mtx);
+}
+
+static void
+nni_posix_poll_thr(void *arg)
+{
+	nni_posix_pollq *pq = arg;
+	struct kevent    kevents[NNI_MAX_KQUEUE_EVENTS];
+
+	nni_mtx_lock(&pq->mtx);
+
+	while (!pq->close) {
+		int i;
+		int nevents;
+
+		// block indefinitely, timers are handled separately
+		nni_mtx_unlock(&pq->mtx);
+		nevents = kevent(
+		    pq->kq, NULL, 0, kevents, NNI_MAX_KQUEUE_EVENTS, NULL);
+		nni_mtx_lock(&pq->mtx);
+
+		if (nevents < 0) {
+			continue;
+		}
+
+		// dispatch events
+		for (i = 0; i < nevents; ++i) {
+			const struct kevent * ev;
+			nni_posix_pollq_node *node;
+
+			ev = &kevents[i];
+
+			if ((int) ev->ident == pq->wakerfd &&
+			    ev->filter == EVFILT_READ) {
+				nni_plat_pipe_clear(pq->wakerfd);
+				continue;
+			}
+
+			node = (nni_posix_pollq_node *) ev->udata;
+			if (node->pq == NULL) {
+				// node was removed while we were blocking
+				continue;
+			}
+			node->revents = 0;
+
+			if (ev->flags & (EV_ERROR | EV_EOF)) {
+				node->revents |= POLLHUP;
+			}
+			if (ev->filter == EVFILT_WRITE) {
+				node->revents |= POLLOUT;
+			} else if (ev->filter == EVFILT_READ) {
+				node->revents |= POLLIN;
+			} else {
+				NNI_ASSERT(false); // unhandled filter
+			}
+
+			// clear events
+			node->events &= ~node->revents;
+
+			// Save the active node; we can notice this way
+			// when it is busy, and avoid freeing it until
+			// we are sure that it is not in use.
+			pq->active = node;
+
+			// Execute the callback with lock released
+			nni_mtx_unlock(&pq->mtx);
+			node->cb(node->data);
+			nni_mtx_lock(&pq->mtx);
+
+			// We finished with this node.  If something
+			// was blocked waiting for that, wake it up.
+			pq->active = NULL;
+			if (pq->wait == node) {
+				pq->wait = NULL;
+				nni_cv_wake(&pq->cv);
+			}
+		}
+	}
+
+	nni_mtx_unlock(&pq->mtx);
+}
+
+static void
+nni_posix_pollq_destroy(nni_posix_pollq *pq)
+{
+	if (pq->started) {
+		nni_mtx_lock(&pq->mtx);
+		pq->close   = 1;
+		pq->started = 0;
+		nni_plat_pipe_raise(pq->wakewfd);
```

Comment:
I think we can probably dispense with the wakewfd and wakerfd and the plat pipe logic.  That logic exists because I don't have any other way to alert a pollq using poll to a change in the polled set of descriptors.  I think kqueue doesn't suffer from this problem at all, since it uses explicit system calls for this.

Reply:
You do need to do something to cause the poll thread to notice and exit though, so maybe this pipe is still useful?  Alternatively what happens if you just close the kqueue fd?  Does the thread wake up?  Is there some other convenient way to force an event via the kqueue system call?

Reply:
closing the kqueue is enough to wake it up and exit - will remove wake fds for now, can always bring them back if needed.